### PR TITLE
#144 Cache scoped items

### PR DESCRIPTION
**Pull request by MaciejWanat**
I've done a bit more than I suggested in https://github.com/pamidur/aspect-injector/issues/144, so here's the list of things I've done:

- I think suggestion with parameter of `PerInstance` is great! The only difference I made is that I believe default behavior should NOT be per instance. If you think it should be otherwise, it will be easy to change
- I've changed way of calculating hash of arguments. Previous way was incorrect, and if you had function accepting two same type arguments, their order didn't matter (see the test).
- `GetKey` now has two overloads - with and without instance. I also changed it to be `protected` instead `private`, so anyone can define their own way of creating a key, in they way they find suitable, by overriding these methods.
- Added tests for following changes
- Fixed some minor things, removed unused code, typos, renamed to match convention

I got a bit confused in line 58, because there are no tests for multiple triggers. So I figured I will use `Any` to define which key to use, but I'm not sure whether its correct.

I hope you like the changes :)

### CODE REVIEW

**Comment by MaciejWanat** [comment]
This failed test is passing when ran directly, doesn't when tests are run altogether.

**Comment by MaciejWanat** [comment]
Using `method.DeclaringType.FullName` ends up with behaving in really strange way, so I've changed it to `method`, which should work just as good - tests passes correctly.

File: samples/src/Cache/CacheAspect.cs [hunk 1]
```diff
@@ -50,16 +54,17 @@ public class CacheAspect
             object result = null;
             var resultFound = false;
 
-            var cacheTriggers = triggers.OfType<CacheAttribute>();
-            var key = GetKey(instance, target.Method, args);
+            var cacheTriggers = (triggers.OfType<CacheAttribute>()).ToList();
+            var useInstance = cacheTriggers.Any(ct => ct.PerInstanceCache);
```
Comment (pamidur) [inline samples/src/Cache/CacheAspect.cs:58]: This then should be calculated on per cacheTrigger inside foreach block.
Reason is - the aspect is basically a processor that executes caching logic for all types of cache. And there could be several types of cache simultaneously, like ```MemoryCache(10, PerIntsanceCache= true)``` and ```MemoryCache(300, PerIntsanceCache= false)``` and probably ```RedisCache(9000, PerIntsanceCache= true) for the same method.

File: samples/src/Cache/CacheAspect.cs [hunk 2]
```diff
@@ -14,29 +14,33 @@ public abstract class CacheAttribute : Attribute
...
         public override CacheItemPolicy Policy => new CacheItemPolicy() { AbsoluteExpiration = DateTimeOffset.UtcNow.AddSeconds(_seconds) };
     }
 
 
     [Aspect(Scope.Global)]
     public class CacheAspect
     {
-        private static readonly Type _voidTaskResult = Task.FromException(new Exception()).GetType();
-        private static readonly object _nullmarker = new object();
+        private static readonly object NullMarker = new object();
```
Comment (pamidur) [inline samples/src/Cache/CacheAspect.cs:43]: Why naming change?

**Review by pamidur — CHANGES_REQUESTED** [review]
Very good PR! Just a few changes requested

File: samples/src/Cache/CacheAspect.cs (hunk 2 above)
Comment (MaciejWanat) [reply samples/src/Cache/CacheAspect.cs:43]: CamelCase is a standard recommended for static fields: https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/d53b55ey(v=vs.71)?redirectedfrom=MSDN

File: samples/src/Cache/CacheAspect.cs (hunk 1 above)
Comment (MaciejWanat) [reply samples/src/Cache/CacheAspect.cs]: I've modified it to be calculated per trigger. It's a bit inefficient that I calculate both keys upfront, but this way it stays readable and easy to override.

File: samples/src/Cache/CacheAspect.cs (hunk 2 above)
Comment (pamidur) [reply samples/src/Cache/CacheAspect.cs:43]: In this project we use CamelCase only for public fields, for private we use underscore prefix. I know, it is my fault I shoud've written the contribution guide long ago. That is a good push for me.

File: samples/src/Cache/CacheAspect.cs (hunk 1 above)
Comment (pamidur) [reply samples/src/Cache/CacheAspect.cs]: I agree, slightly inefficient, but what are our options?

**Review by pamidur — APPROVED** [review]

**Comment by MaciejWanat** [comment]
@pamidur when is this change expected to be realeased?

**Comment by pamidur** [comment]
Monday-Tuesday I think, I have a few more fixes in mind.

**Comment by pamidur** [comment]
@MaciejWanat, one thing I will change is PerInstance by default equals true, to be backward compatible with previous versions

**Comment by MaciejWanat** [comment]
@pamidur good idea, I was thinking about the same suggestion

**Comment by pamidur** [comment]
I also found an issue in AI with properties from base classes in attributes, I need to fix before the release

**Comment by pamidur** [comment]
Released as https://www.nuget.org/packages/Aspects.Cache/1.0.2

**Pull request merged**